feat(pool_config): dump all pools config in json format

vishnuitta opened the pull request:
Signed-off-by: Vitta <vitta@mayadata.io>

<!--- Provide a general summary of your changes in the Title above -->
<!--- Explain how the fix was tested -->

This PR prints the config of all pools in json format.

vishnuitta committed:
feat(pool_config): dump all pools config in json format

Signed-off-by: Vitta <vitta@mayadata.io>

mynktl commented on module/nvpair/nvpair_json.c:
```diff
@@ -0,0 +1,508 @@
+ * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017 by Delphix. All rights reserved.
+ */
+
```
If we have modified this file then let's append our openebs license here.

mynktl commented on module/nvpair/nvpair.c:
```diff
@@ -3334,6 +3334,9 @@ MODULE_AUTHOR(ZFS_META_AUTHOR);
 MODULE_LICENSE(ZFS_META_LICENSE);
 MODULE_VERSION(ZFS_META_VERSION "-" ZFS_META_RELEASE);
 
+EXPORT_SYMBOL(nvlist_dump_json);
```
No need for `EXPORT_SYMBOL` at line number 3337 and 3338, since we are not doing user-space compilation for this file.

mynktl replied:
Ignore this.

mynktl reviewed (dismissed):
Changes are good.

mynktl commented on lib/libnvpair/Makefile.am:
```diff
@@ -14,6 +14,7 @@ lib_LTLIBRARIES = libnvpair.la
 
 USER_C = \
 	libnvpair.c \
+	nvpair_json.c \
```
`nvpair_json.c` should be under `KERNEL_C`, right?

vishnuitta committed:
fix(travis): compile issue with kernel build

Signed-off-by: Vitta <vitta@mayadata.io>

vishnuitta committed:
fix(license): header with only copyright

Signed-off-by: Vitta <vitta@mayadata.io>

vishnuitta committed:
fix(license): removed extra comment section

Signed-off-by: Vitta <vitta@mayadata.io>

vishnuitta committed:
fix(new_file): moved new file location as per original file location

Signed-off-by: Vitta <vitta@mayadata.io>

pawanpraka1 commented on cmd/zpool/zpool_main.c:
```diff
@@ -6556,6 +6561,46 @@ status_callback(zpool_handle_t *zhp, void *data)
+{
+	int ret;
+	status_cbdata_t cb = { 0 };
+	cb.cb_allpools = B_TRUE;
```
both the lines are not needed.

vishnuitta replied:
we need to set cb_allpools to iterator to do callback for all pools

vishnuitta committed:
fix(travis): compilation issue in file location

Signed-off-by: Vitta <vitta@mayadata.io>

vishnuitta committed:
fix(travis): delete unnecessary extra file

Signed-off-by: Vitta <vitta@mayadata.io>

vishnuitta committed:
fix(travis): having zpool dump only in UZFS mode

Signed-off-by: Vitta <vitta@mayadata.io>

vishnuitta committed:
fix(travis): export symbols only in uzfs mode

Signed-off-by: Vitta <vitta@mayadata.io>

vishnuitta committed:
fix(exports): remove unnecessary exported functions

Signed-off-by: Vitta <vitta@mayadata.io>

vishnuitta committed:
travis(fix): compilation issue

Signed-off-by: Vitta <vitta@mayadata.io>

vishnuitta commented:
this PR is ready for review

richardelling commented on cmd/zpool/zpool_main.c:
```diff
@@ -6556,6 +6565,48 @@ status_callback(zpool_handle_t *zhp, void *data)
 	return (0);
 }
 
+#ifdef _UZFS
```
I don't think we need this ifdef, the code will work properly for in-kernel ZFS, too.

richardelling commented on cmd/zpool/zpool_main.c:
```diff
@@ -6556,6 +6565,48 @@ status_callback(zpool_handle_t *zhp, void *data)
+#ifdef _UZFS
+/*
+ * Dumps pool's config in json
+ */
```
Can we add a comment block here:
```
* The current output is the raw configuration nvlist converted to JSON.
* Not all of the dumped values are represented by keys, some are arrays
* that represent internal structs. Hence their contents are not considered
* stable across different versions.
* TODO: convert struct arrays into key:value pairs
```

richardelling commented on module/nvpair/nvpair_json.c:
```diff
@@ -0,0 +1,512 @@
+/*
+ * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017 by Delphix. All rights reserved.
+ * Copyright © 2017-2019 The OpenEBS Authors
```
we try to keep unicode out of the source, this might get flagged by static style checkers

richardelling commented on cmd/zpool/zpool_main.c:
```diff
@@ -347,6 +354,8 @@ get_usage(zpool_help_t idx)
 	case HELP_STATUS:
 		return (gettext("\tstatus [-c [script1,script2,...]] [-gLPvxD]"
 		    "[-T d|u] [pool] ... [interval [count]]\n"));
+	case HELP_DUMP:
```
missing ifdef?  see also comment below

richardelling commented on cmd/zpool/zpool_main.c:
```diff
@@ -142,6 +145,7 @@ typedef enum {
 	HELP_REMOVE,
 	HELP_SCRUB,
 	HELP_STATUS,
+	HELP_DUMP,
```
missing ifdef? see also comment below

richardelling reviewed (commented):
To be complete, updating the zpool man page would be helpful.
Also, I added a TODO in the comments. I have some code that can be
contributed to address some of the TODO items, but they can be done
later, if that helps the schedule.

vishnuitta committed:
fix(reviews): incorporate comments to remove unnecessary macros

Signed-off-by: Vitta <vitta@mayadata.io>

vishnuitta committed:
fix(license): removed unicode character from license line

Signed-off-by: Vitta <vitta@mayadata.io>

vishnuitta commented:
> To be complete, updating the zpool man page would be helpful.
> Also, I added a TODO in the comments. I have some code that can be
> contributed to address some of the TODO items, but they can be done
> later, if that helps the schedule.

Thanks for the comments. Addressed the review comments.

vishnuitta commented:
@richardelling this PR is open for reviews and can be taken in next PR if any reviews are there.
Merging the PR as all review comments are addressed and also this PR is good wrt intended functionality.
